raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -3352,8 +3364,13 @@ private boolean shouldSendAddOrRemoveVoterRequest(FollowerState state, long curr
          * and are configured to auto join should attempt to automatically join the voter
          * set for the configured topic partition.
          */
-        return partitionState.lastKraftVersion().isReconfigSupported() && canBecomeVoter &&
-            quorumConfig.autoJoin() && state.hasUpdateVoterSetPeriodExpired(currentTimeMs);
+        if (!partitionState.lastKraftVersion().isReconfigSupported() || !canBecomeVoter ||
+            !quorumConfig.autoJoin() || !state.hasUpdateVoterSetPeriodExpired(currentTimeMs)) {
+            return false;
+        }
+
+        // Only attempt auto-join if we haven't already auto-joined
+        return !hasAutoJoined;
```

Comment:
```suggestion
        return !hasAutoJoined && partitionState.lastKraftVersion().isReconfigSupported() && canBecomeVoter && quorumConfig.autoJoin() && state.hasUpdateVoterSetPeriodExpired(currentTimeMs);
```
IMO this reads cleaner. It will also short circuit after a node successfully auto-joins and sets `hasAutoJoined = true`, whereas this code evaluates all boolean expressions always.